File: problems/2025/09-20/longest-subarray-length/solution.js
class Solution {
  longestSubarray(arr) {
    const n = arr.length;
    const left = Array(n).fill(-1);
    const right = Array(n).fill(n);
    const st = [];
    for (let i = 0; i < n; i++) {
      while (st.length && arr[st[st.length - 1]] <= arr[i]) st.pop();
      left[i] = st.length ? st[st.length - 1] : -1;
      st.push(i);
    }
    st.length = 0;
    for (let i = n - 1; i >= 0; i--) {
      while (st.length && arr[st[st.length - 1]] < arr[i]) st.pop();
      right[i] = st.length ? st[st.length - 1] : n;
      st.push(i);
    }
    let ans = 0;
    for (let i = 0; i < n; i++) {
      const span = right[i] - left[i] - 1;
      if (span >= arr[i]) ans = Math.max(ans, span);
    }
    return ans;
  }
}

const sol = new Solution();
console.log(sol.longestSubarray([1,2,3]));     // 3
console.log(sol.longestSubarray([6,4,2,5]));   // 0

File: problems/2025/09-20/longest-subarray-length/Solution.java
import java.util.*;
public class Solution {
    public static int longestSubarray(int[] arr) {
        int n = arr.length;
        int[] left = new int[n];
        int[] right = new int[n];
        Arrays.fill(left, -1);
        Arrays.fill(right, n);
        Deque<Integer> st = new ArrayDeque<>();
        // left: nearest greater strictly to the left
        for (int i = 0; i < n; i++) {
            while (!st.isEmpty() && arr[st.peek()] <= arr[i]) st.pop();
            left[i] = st.isEmpty() ? -1 : st.peek();
            st.push(i);
        }
        st.clear();
        // right: nearest greater strictly to the right
        for (int i = n - 1; i >= 0; i--) {
            while (!st.isEmpty() && arr[st.peek()] < arr[i]) st.pop();
            right[i] = st.isEmpty() ? n : st.peek();
            st.push(i);
        }
        int ans = 0;
        for (int i = 0; i < n; i++) {
            int span = right[i] - left[i] - 1;
            if (span >= arr[i]) ans = Math.max(ans, span);
        }
        return ans;
    }
    public static void main(String[] args) {
        System.out.println(longestSubarray(new int[]{1,2,3}));    // 3
        System.out.println(longestSubarray(new int[]{6,4,2,5}));  // 0
    }
}
